### ISSUE TITLE
onPause() doesn't stop pinging

### ISSUE THREAD

**Issue by mudar** [mapbox/mapbox-gl-native#3667]
Using android lib, version `4.0.0-SNAPSHOT`
When leaving the map activity, I call both `map.onPause()` and `map.onStop()`. However, the mapview seems to keep pinging mapbox servers 4-6 times / second. 
I see the following in the log
```
D/com.mapbox.mapboxsdk.http.HTTPContext$HTTPRequest: [HTTP] Request with response code = 304: Not Modified
```
And it doesn't stop till I manually kill my app

**Comment by jfirebaugh** [comment]
cc @kkaefer

**Comment by tobrun** [comment]
FWIW I'm also seeing this on an idle map 
eg. open `MainActivity` of testapp and do nothing
this results in a flood of `304: not modified messages` (14 request/sec).

**Comment by mayardb** [comment]
I am reproducing the same issue as @tobrun.

( + ANR when leaving the MapView, but don't know if it is related)

**Comment by kkaefer** [comment]
@mudar What style are you using? Mapbox GL should definitely not ping the server repeatedly, even when not paused/stopped. It seems like this is a bug with HTTP request revalidation, prompting it to consider the resource stale.

**Comment by mudar** [comment]
I'm using a custom style.
I can try testing again with a plain one but it doesn't happen all the time, so not sure how that would go..

When I noticed the logs, I had left my app (running in the background), so that did seem strange.

**Comment by tobrun** [comment]
@kkaefer and me have been trying to figure this one out. We noticed that @kkaefer was not able to reproduce this locally with current master on mulitple devices while I was having this constantly with a build from Friday. After I updated my build locally the problem disappeared. This means that when a new SNAPSHOT lands the problem should be gone as well. 

@mudar 
could you test this with the new SNAPSHOT from tomorrow? 
Reopen if this message is still shown and thank you again for reaching out and reporting this problem!

**Comment by mudar** [comment]
I was running this on Nexus 5 running version 6.0

**Comment by kkaefer** [comment]
@mudar What version/build number of Mapbox GL were you using?

**Comment by mudar** [comment]
4.0.0-SNAPSHOT, using gradle

I have since replaced that build on my device. So if the snapshot was updated since (from previous comments, it seems that's the case), then I guess we can try guessing the build number from the issue's timestamp: `Jan 22nd, 13:10 EST`

**Comment by kkaefer** [comment]
Did you try with a current build as well?

**Comment by jfirebaugh** [comment]
Also pertinent is what the `Expires` and `Cache-Control` response headers are for the request(s) that are being revalidated frequently.

**Comment by brewin** [comment]
I'm seeing the same thing. The raster tile layer I'm using doesn't have Expires or Cache-Control set. Is there a way to disable automatic revalidation?

**Comment by tobrun** [comment]
@mudar @mayardb @brewin :
Would you be able to retest this with the new `SNAPSHOT`?
Be sure to refresh your dependency by executing:
> gradle --refresh-dependencies INSERT_TASK_HERE

**Comment by mayardb** [comment]
@tobrun :
Still reproducing.

Here are my logs when leaving:
```
01-26 11:53:26.694 I/mbgl: {Map}[Android]: Not activating as we are not ready
01-26 11:53:26.694 D/com.mapbox.mapboxsdk.http.HTTPContext$HTTPRequest: [HTTP] Request with response code = 304: Not Modified
01-26 11:53:26.714 D/com.mapbox.mapboxsdk.http.HTTPContext$HTTPRequest: [HTTP] Request with response code = 304: Not Modified
01-26 11:53:26.734 W/com.mapbox.mapboxsdk.http.HTTPContext$HTTPRequest: [HTTP] Request could not be executed: Canceled
01-26 11:53:26.734 W/com.mapbox.mapboxsdk.http.HTTPContext$HTTPRequest: [HTTP] Request could not be executed: Canceled
01-26 11:53:26.734 E/libEGL: call to OpenGL ES API with no current context (logged once per thread)
01-26 11:53:26.734 I/mbgl: {Map}[Android]: Not deactivating as we are not ready
```

**Comment by tobrun** [comment]
@mayardb 
It is normal to see some `Not Modified` since these calls happen async but does it keep executing these?

**Comment by mayardb** [comment]
While the application is running I still see these requests in the logcat.
Then when I am leaving the application freeze.

(By the way, I am not in the test app context, but in my own app using mapbox)

**Comment by kkaefer** [comment]
@mayardb to clarify what @tobrun said, it's expected to so *some* 304 messages, in particular during app startup or when the user interacts with the map. What you shouldn't see is continued 304 messages when you're not interacting with the map, in particular not multiple 304 messages per second.

**Comment by brewin** [comment]
Using the latest snapshot, Mapbox GL still hammers my server with requests. My tile server didn't have Expires or Cache-Control set. When I set Expires to '-1' or 'off' and Cache-Control to 'no-cache', there is still a flood of 304s. When I set Expires to 1 second, the requests stop.

**Comment by tobrun** [comment]
Going to add a Activity stack feature to the Testapp and see if I'm able to reproduce this. 
The feature will cycle through Activities without or without a Map.
I believe this resembles a typical use-case for end-developers that is testable with espresso.

**Comment by mudar** [comment]
I cannot reproduce this issue anymore, probably because I'm using SNAPSHOT. Sorry, but I guess that's good news!
So if anyone else runs into this issue please report the mapbox-library build number here.

**Comment by jfirebaugh** [comment]
There are two issues here:

* Pausing the map doesn't pause resource revalidation. This is a legitimate, but non-critical, issue.
* Resources without cache headers may be endlessly revalidated (regardless of pause state). This is caused by 1c21d0fd4cd30cbf6c5b863fd0179b227c28bc0b. Prior to that commit, resources without an expiration were stored with an `expiration` column value of 0. After it, they are stored with a null column value, but existing resources in the cache with a 0 value are treated as expiring at the epoch. We've bumped the cache schema recently, which has the effect of invalidating existing caches. That may make this an issue only for people who used SNAPSHOT builds. But this needs to verified, and even if it's the case we may want to bump the schema version again just to be safe.

**Comment by jfirebaugh** [comment]
@bleege This _is_ going to be an issue for 3.2.0. The last schema version bump was released in 3.1.0, and 1c21d0fd4cd30cbf6c5b863fd0179b227c28bc0b is in 3.2.0. We'll need to make a 3.2.1 release ASAP.

**Comment by bleege** [comment]
Discussed with @jfirebaugh and confirmed that this issue only impacts 3rd Party Tile Servers and NOT tiles served from Mapbox.  The fix for this is in PR #3733 which we'll keep on track for `Android 4.0.0`.

**Comment by mudar** [comment]
This sounds strange, our custom style is served from mapbox
`mapbox://styles/arnaudspuhler/cigfc11g0000dcnm73r4o3pi1`

**Comment by tobrun** [comment]
Haven't seen this or heard user reporting this issue. Closing for now, reopen if this is still an issue.

**Comment by jfirebaugh** [comment]
Possible causes of frequent revalidation were fixed in #3733 and #4019.

**Comment by s3r6** [comment]
I freauently get such logs on 4.2.0 beta Android SDK. How can I get detallied info about the URL and other parameters of the request?

`W/com.mapbox.mapboxsdk.http.HTTPRequest: [HTTP] Request could not be executed: Socket closed`
